--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -54,7 +_,16 @@
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
 
-public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
+public class Connection extends SimpleChannelInboundHandler<Packet<?>> implements spout.server.paper.impl.clientview.lookup.ClientViewLookup { // Spout - Client view - Lookup - Implement for Connection
+    public spout.server.paper.api.clientview.@Nullable ClientView clientView; // Spout - Client view - Set - Field
+    // Spout start - Client view - Lookup - Implement for Connection - Getter
+    @Override
+    public spout.server.paper.api.clientview.@Nullable ClientView getClientView() {
+        return this.clientView;
+    }
+    // Spout end - Client view - Lookup - Implement for Connection - Getter
+    public final spout.server.paper.impl.clientview.set.ClientViewSetter clientViewSetter; // Spout - Client view - Set - Setter
+    public final spout.server.paper.impl.packetmapping.item.reverse.ItemMappingReverser itemMappingReverser; // Spout - Item mapping - Reverse - Store reverser in Connection
     private static final float AVERAGE_PACKETS_SMOOTHING = 0.75F;
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final Marker ROOT_MARKER = MarkerFactory.getMarker("NETWORK");
@@ -130,12 +_,15 @@
 
     public Connection(final PacketFlow receiving) {
         this.receiving = receiving;
+        this.clientViewSetter = new spout.server.paper.impl.clientview.set.ClientViewSetter(this); // Spout - Client view - Set - Setter
+        this.itemMappingReverser = new spout.server.paper.impl.packetmapping.item.reverse.ItemMappingReverser(); // Spout - Item mapping - Reverse - Store reverser in Connection
     }
 
     @Override
     public void channelActive(final ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
         this.channel = ctx.channel();
+        this.channel.attr(spout.server.paper.impl.clientview.lookup.packethandling.NettyClientViewLookupAttribute.ATTRIBUTE_KEY).set(new java.lang.ref.WeakReference<>(this)); // Spout - Client view - Lookup - During packet handling - Copy the connection (acting as the lookup) to the channel
         this.address = this.channel.remoteAddress();
         this.preparing = false; // Spigot
         if (this.delayedDisconnect != null) {
