--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -67,10 +_,21 @@
 import org.joml.Vector3f;
 import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
+import spout.api.clientview.model.ClientView;
 
-public class FriendlyByteBuf extends ByteBuf {
+public class FriendlyByteBuf extends ByteBuf implements spout.server.paper.impl.clientview.lookup.ClientViewLookup { // Spout - Client view - Lookup - During packet handling - Store provider in FriendlyByteBuf - Easy access via ClientViewProvider interface
     private final ByteBuf source;
     public final java.util.@Nullable Locale adventure$locale; // Paper - track player's locale for server-side translations
+    public final java.lang.ref.@Nullable WeakReference<spout.server.paper.impl.clientview.lookup.ClientViewLookup> clientViewProvider; // Spout - Client view - Lookup - During packet handling - Store provider in FriendlyByteBuf
+    // Spout start - Client view - Lookup - During packet handling - Store provider in FriendlyByteBuf - Easy access via ClientViewProvider interface
+    @Override
+    public @Nullable ClientView getClientView() {
+        spout.server.paper.impl.clientview.lookup.@Nullable ClientViewLookup clientViewLookup = this.clientViewProvider == null ? null : this.clientViewProvider.get();
+        return clientViewLookup == null ? null : clientViewLookup.getClientView();
+    }
+    // Spout end - Client view - Lookup - During packet handling - Store provider in FriendlyByteBuf - Easy access via ClientViewProvider interface
+    public boolean isEncodingStonecutterRecipeResult; // Spout - Item mapping - Detect stonecutter result - Store in FriendlyByteBuf
+    public boolean isEncodingItemFrameItem; // Spout - Item mapping - Detect item frame - Store in FriendlyByteBuf
     public static final short MAX_STRING_LENGTH = 32767;
     public static final int MAX_COMPONENT_STRING_LENGTH = 262144;
     private static final int PUBLIC_KEY_SIZE = 256;
@@ -84,6 +_,7 @@
 
     public FriendlyByteBuf(final ByteBuf source) {
         this.adventure$locale = PacketEncoder.ADVENTURE_LOCALE.get(); // Paper - track player's locale for server-side translations
+        this.clientViewProvider = spout.server.paper.impl.clientview.lookup.packethandling.ClientViewLookupThreadLocal.THREAD_LOCAL.get();// Spout - Client view - Lookup - During packet handling - Copy lookup from the thread-local variable to the write buffer
         this.source = source;
     }
 
