--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -669,11 +_,14 @@
 
     @Override
     public void onLightUpdate(final LightLayer layer, final SectionPos pos) {
+        boolean requiresExtraLightSending = spout.server.paper.impl.packetmapping.block.light.CurrentTaskRequiresExtraLightSending.get(); // Spout - Block mapping - Update block light for custom blocks
         this.mainThreadProcessor.execute(() -> {
             ChunkHolder chunk = this.getVisibleChunkIfPresent(pos.chunk().pack());
+            if (requiresExtraLightSending) spout.server.paper.impl.packetmapping.block.light.CurrentBroadcastRequiresExtraLightSending.set(); // Spout - Block mapping - Update block light for custom blocks
             if (chunk != null && chunk.sectionLightChanged(layer, pos.y())) {
                 this.chunkHoldersToBroadcast.add(chunk);
             }
+            if (requiresExtraLightSending) spout.server.paper.impl.packetmapping.block.light.CurrentBroadcastRequiresExtraLightSending.clear(); // Spout - Block mapping - Update block light for custom blocks
         });
     }
 
