--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -38,10 +_,12 @@
     protected final ServerPlayer player;
     private GameType gameModeForPlayer = GameType.DEFAULT_MODE;
     private @Nullable GameType previousGameModeForPlayer;
-    private boolean isDestroyingBlock;
-    private int destroyProgressStart;
+    // Spout start - Block mapping - Breaking speed - Keep ratio history for compensation - Make fields public
+    public boolean isDestroyingBlock;
+    public int destroyProgressStart;
     private BlockPos destroyPos = BlockPos.ZERO;
-    private int gameTicks;
+    public int gameTicks;
+    // Spout end - Block mapping - Breaking speed - Keep ratio history for compensation - Make fields public
     private boolean hasDelayedDestroy;
     private BlockPos delayedDestroyPos = BlockPos.ZERO;
     private int delayedTickStart;
@@ -283,6 +_,7 @@
                     BlockState state = this.level.getBlockState(pos);
                     if (!state.isAir()) {
                         float destroyProgress = state.getDestroyProgress(this.player, this.player.level(), pos) * (ticksSpentDestroying + 1);
+                        destroyProgress *= spout.server.paper.impl.packetmapping.block.breakspeed.BlockBreakSpeedLatencyCompensation.getCompensationFactor(this.player, this.player.level(), pos, state); // Spout - Block mapping - Breaking speed - Compensate for difference in intended ratio and recently observed ratio
                         if (destroyProgress >= 0.7F) {
                             this.isDestroyingBlock = false;
                             this.level.destroyBlockProgress(this.player.getId(), pos, -1);
