--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -321,7 +_,7 @@
             // CraftBukkit end
         }
 
-        Packet<ClientGamePacketListener> packet = this.entity.getAddEntityPacket(this);
+        Packet<ClientGamePacketListener> packet = this.entity.getAddEntityPacket(this, player); // Spout - Packet mapping - Implementation part - Apply to entities
         broadcast.accept(packet);
         if (this.trackedDataValues != null) {
             broadcast.accept(new ClientboundSetEntityDataPacket(this.entity.getId(), this.trackedDataValues));
@@ -330,12 +_,14 @@
         if (this.entity instanceof LivingEntity livingEntity) {
             Collection<AttributeInstance> attributes = livingEntity.getAttributes().getSyncableAttributes();
             // CraftBukkit start - If sending own attributes send scaled health instead of current maximum health
+            @Nullable ServerPlayer selfPlayer = null; // Spout - Block mapping - Breaking speed - Pass self player to attributes packet
             if (this.entity.getId() == player.getId()) {
+                selfPlayer = player; // Spout - Block mapping - Breaking speed - Pass self player to attributes packet
                 ((ServerPlayer) this.entity).getBukkitEntity().injectScaledMaxHealth(attributes, false);
             }
             // CraftBukkit end
             if (!attributes.isEmpty()) {
-                broadcast.accept(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributes));
+                broadcast.accept(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributes, selfPlayer)); // Spout - Block mapping - Breaking speed - Pass self player to attributes packet
             }
         }
 
@@ -404,7 +_,12 @@
                     serverPlayer.getBukkitEntity().injectScaledMaxHealth(attributes, false);
                 }
                 // CraftBukkit end
-                this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributes));
+                // Spout start - Block mapping - Breaking speed - Pass self player to attributes packet
+                this.synchronizer.sendToTrackingPlayers(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributes, null));
+                if (this.entity instanceof ServerPlayer serverPlayer) {
+                    serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket(this.entity.getId(), attributes, serverPlayer));
+                }
+                // Spout ebd - Block mapping - Breaking speed - Pass self player to attributes packet
             }
 
             attributes.clear();
