--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -46,11 +_,11 @@
 import net.minecraft.world.level.material.PushReaction;
 
 public class Blocks {
-    private static final BlockBehaviour.StatePredicate NOT_CLOSED_SHULKER = (statex, level, pos) -> !(
+    public static final BlockBehaviour.StatePredicate NOT_CLOSED_SHULKER = (state, level, pos) -> !( // Spout - Common class tweakers - Make accessible
             level.getBlockEntity(pos) instanceof ShulkerBoxBlockEntity shulkerBoxBlockEntity
         )
         || shulkerBoxBlockEntity.isClosed();
-    private static final BlockBehaviour.StatePredicate NOT_EXTENDED_PISTON = (statex, level, pos) -> !statex.getValue(PistonBaseBlock.EXTENDED);
+    public static final BlockBehaviour.StatePredicate NOT_EXTENDED_PISTON = (state, level, pos) -> !state.getValue(PistonBaseBlock.EXTENDED); // Spout - Common class tweakers - Make accessible
     public static final Block AIR = register("air", AirBlock::new, BlockBehaviour.Properties.of().replaceable().noCollision().noLootTable().air());
     public static final Block STONE = register(
         "stone",
@@ -2062,7 +_,7 @@
             .strength(0.1F)
             .requiresCorrectToolForDrops()
             .sound(SoundType.SNOW)
-            .isViewBlocking((statex, level, pos) -> statex.getValue(SnowLayerBlock.LAYERS) >= 8)
+            .isViewBlocking(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.MAX_SNOW_LAYERS) // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
             .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ICE = register(
@@ -6226,7 +_,7 @@
             .strength(1.5F)
             .sound(SoundType.SCULK_SENSOR)
             .lightLevel(statex -> 1)
-            .emissiveRendering((statex, level, pos) -> SculkSensorBlock.getPhase(statex) == SculkSensorPhase.ACTIVE)
+            .emissiveRendering(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.SCULK_PHASE_ACTIVE) // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
     );
     public static final Block CALIBRATED_SCULK_SENSOR = register(
         "calibrated_sculk_sensor", CalibratedSculkSensorBlock::new, BlockBehaviour.Properties.ofLegacyCopy(SCULK_SENSOR)
@@ -7045,7 +_,7 @@
             .pushReaction(PushReaction.DESTROY)
     );
 
-    private static ToIntFunction<BlockState> litBlockEmission(final int lightEmission) {
+    private static ToIntFunction<net.minecraft.world.level.block.state.BlockState> litBlockEmission(final int lightEmission) {
         return state -> state.getValue(BlockStateProperties.LIT) ? lightEmission : 0;
     }
 
@@ -7053,7 +_,7 @@
         return blockState -> blockState.getValue(BlockStateProperties.WATERLOGGED) ? MapColor.WATER : mapColor;
     }
 
-    private static Boolean never(final BlockState state, final BlockGetter blockGetter, final BlockPos blockPos, final EntityType<?> entityType) {
+    public static Boolean never(BlockState state, BlockGetter level, BlockPos pos, EntityType<?> entity) { // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Defaults for types - Make public
         return false;
     }
 
@@ -7061,7 +_,7 @@
         return true;
     }
 
-    private static Boolean ocelotOrParrot(final BlockState state, final BlockGetter blockGetter, final BlockPos blockPos, final EntityType<?> entityType) {
+    public static Boolean ocelotOrParrot(final BlockState state, final BlockGetter blockGetter, final BlockPos blockPos, final EntityType<?> entityType) { // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Defaults for types - Make public
         return entityType == EntityType.OCELOT || entityType == EntityType.PARROT;
     }
 
@@ -7089,14 +_,14 @@
     }
 
     private static BlockBehaviour.Properties netherStemProperties(final MapColor mapColor) {
-        return BlockBehaviour.Properties.of().mapColor(state -> mapColor).instrument(NoteBlockInstrument.BASS).strength(2.0F).sound(SoundType.STEM);
+        return BlockBehaviour.Properties.of().mapColor(mapColor).instrument(NoteBlockInstrument.BASS).strength(2.0F).sound(SoundType.STEM); // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
     }
 
-    private static boolean always(final BlockState state, final BlockGetter blockGetter, final BlockPos blockPos) {
+    public static boolean always(BlockState state, BlockGetter level, BlockPos pos) { // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable - Make public
         return true;
     }
 
-    private static boolean never(final BlockState state, final BlockGetter blockGetter, final BlockPos blockPos) {
+    public static boolean never(BlockState state, BlockGetter level, BlockPos pos) { // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable - Make public
         return false;
     }
 
@@ -7202,6 +_,13 @@
         final ResourceKey<Block> id, final Function<BlockBehaviour.Properties, Block> factory, final BlockBehaviour.Properties properties
     ) {
         Block block = factory.apply(properties.setId(id));
+        // Spout start - More data-driven - Vanilla flag - Implementation part
+        block.markAsVanilla();
+        // Currently, there is no way to modify block states of vanilla blocks, so we assume all block states of vanilla blocks are vanilla too
+        for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+            blockState.markAsVanilla();
+        }
+        // Spout end - More data-driven - Vanilla flag - Implementation part
         return Registry.register(BuiltInRegistries.BLOCK, id, block);
     }
 
@@ -7224,9 +_,10 @@
     static {
         for (Block block : BuiltInRegistries.BLOCK) {
             for (BlockState state : block.getStateDefinition().getPossibleStates()) {
-                Block.BLOCK_STATE_REGISTRY.add(state);
+                // Block.BLOCK_STATE_REGISTRY.add(blockState); // Spout - More data-driven - Minecraft registries - Register block states in BlockRegistry#register
                 state.initCache();
             }
         }
+        spout.server.paper.impl.moredatadriven.minecraft.BlockRegistry.get().canInitializeBlockStateCaches = true; // Spout - More data-driven - Minecraft registries - Initialize block state caches for non-vanilla blocks in BlockRegistry#register
     }
 }
