--- a/net/minecraft/world/level/block/MultifaceBlock.java
+++ b/net/minecraft/world/level/block/MultifaceBlock.java
@@ -242,7 +_,7 @@
         return newState;
     }
 
-    public static boolean hasFace(final BlockState state, final Direction faceDirection) {
+    public static boolean hasFace(BlockState state, Direction faceDirection) { // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state function properties serializable
         BooleanProperty property = getFaceProperty(faceDirection);
         return state.getValueOrElse(property, false);
     }
