--- a/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -42,6 +_,7 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
+import spout.common.moredatadriven.minecraft.common.subtypes.BlockStateFunction;
 
 public class RespawnAnchorBlock extends Block {
     public static final MapCodec<RespawnAnchorBlock> CODEC = simpleCodec(RespawnAnchorBlock::new);
@@ -219,7 +_,7 @@
         return true;
     }
 
-    public static int getScaledChargeLevel(final BlockState state, final int maximum) {
+    public static int getScaledChargeLevel(BlockState state, int maximum) {
         return Mth.floor((state.getValue(CHARGE) - 0) / 4.0F * maximum);
     }
 
