--- a/net/minecraft/world/level/block/SeaPickleBlock.java
+++ b/net/minecraft/world/level/block/SeaPickleBlock.java
@@ -23,6 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
+import spout.common.moredatadriven.minecraft.common.subtypes.BlockStateFunction;
 
 public class SeaPickleBlock extends VegetationBlock implements SimpleWaterloggedBlock, BonemealableBlock {
     public static final MapCodec<SeaPickleBlock> CODEC = simpleCodec(SeaPickleBlock::new);
@@ -56,7 +_,7 @@
         }
     }
 
-    public static boolean isDead(final BlockState state) {
+    public static boolean isDead(BlockState state) {
         return !state.getValue(WATERLOGGED);
     }
 
