--- a/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
+++ b/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
@@ -9,14 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class WeatheringCopperStairBlock extends StairBlock implements WeatheringCopper {
-    public static final MapCodec<WeatheringCopperStairBlock> CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(
-                WeatheringCopper.WeatherState.CODEC.fieldOf("weathering_state").forGetter(ChangeOverTimeBlock::getAge),
-                BlockState.CODEC.fieldOf("base_state").forGetter(b -> b.baseState),
-                propertiesCodec()
-            )
-            .apply(i, WeatheringCopperStairBlock::new)
-    );
+    public static final MapCodec<WeatheringCopperStairBlock> CODEC = spout.gamecontent.datadriven.block.BlockCodecs.weatheringCopperStairCodec(WeatheringCopperStairBlock::new); // Spout - More data-driven - Minecraft registries - Block - Use formatted block state codec
     private final WeatheringCopper.WeatherState weatherState;
 
     @Override
