--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -30,7 +_,8 @@
     );
     private static final int TWO_MEGABYTES = 2097152; // Paper - diff on change - if this changes, update PacketEncoder
     private final Map<Heightmap.Types, long[]> heightmaps;
-    private final byte[] buffer;
+    public byte[] buffer; // Spout - Block mapping - Chunk packets - Expose buffer
+    public int bufferLength; // Spout - Block mapping - Chunk packets - Make buffer length explicit
     private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
     // Paper start - Handle oversized block entities in chunks
     private final java.util.List<net.minecraft.network.protocol.Packet<?>> extraPackets = new java.util.ArrayList<>();
@@ -53,6 +_,7 @@
             .filter(entryx -> ((Heightmap.Types)entryx.getKey()).sendToClient())
             .collect(Collectors.toMap(Entry::getKey, entryx -> (long[])((Heightmap)entryx.getValue()).getRawData().clone()));
         this.buffer = new byte[calculateChunkSize(levelChunk)];
+        this.bufferLength = this.buffer.length; // Spout - Block mapping - Chunk packets - Make buffer length explicit
         // Paper start - Anti-Xray - Add chunk packet info
         if (chunkPacketInfo != null) {
             chunkPacketInfo.setBuffer(this.buffer);
@@ -83,14 +_,17 @@
         }
 
         this.buffer = new byte[size];
+        this.bufferLength = this.buffer.length; // Spout - Block mapping - Chunk packets - Make buffer length explicit
         input.readBytes(this.buffer);
         this.blockEntitiesData = ClientboundLevelChunkPacketData.BlockEntityInfo.LIST_STREAM_CODEC.decode(input);
     }
 
     public void write(final RegistryFriendlyByteBuf output) {
         HEIGHTMAPS_STREAM_CODEC.encode(output, this.heightmaps);
-        output.writeVarInt(this.buffer.length);
-        output.writeBytes(this.buffer);
+        // Spout start - Block mapping - Chunk packets - Make buffer length explicit
+        output.writeVarInt(this.bufferLength);
+        output.writeBytes(this.buffer, 0, this.bufferLength);
+        // Spout end - Block mapping - Chunk packets - Make buffer length explicit
         ClientboundLevelChunkPacketData.BlockEntityInfo.LIST_STREAM_CODEC.encode(output, this.blockEntitiesData);
     }
 
