--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -168,8 +_,8 @@
 
 public class BuiltInRegistries {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private static final Map<Identifier, Supplier<?>> LOADERS = Maps.newLinkedHashMap();
-    private static final WritableRegistry<WritableRegistry<?>> WRITABLE_REGISTRY = new MappedRegistry<>(
+    public static final Map<Identifier, Supplier<?>> LOADERS = Maps.newLinkedHashMap(); // Spout - More data-driven - Minecraft registries - Make public
+    public static final WritableRegistry<WritableRegistry<?>> WRITABLE_REGISTRY = new MappedRegistry<>( // Spout - More data-driven - Minecraft registries - Make public
         ResourceKey.createRegistryKey(Registries.ROOT_REGISTRY_NAME), Lifecycle.stable()
     );
     public static final DataComponentInitializers DATA_COMPONENT_INITIALIZERS = new DataComponentInitializers();
@@ -177,12 +_,12 @@
     public static final Registry<SoundEvent> SOUND_EVENT = registerSimple(Registries.SOUND_EVENT, registry -> SoundEvents.ITEM_PICKUP);
     public static final DefaultedRegistry<Fluid> FLUID = registerDefaultedWithIntrusiveHolders(Registries.FLUID, "empty", registry -> Fluids.EMPTY);
     public static final Registry<MobEffect> MOB_EFFECT = registerSimple(Registries.MOB_EFFECT, MobEffects::bootstrap);
-    public static final DefaultedRegistry<Block> BLOCK = registerDefaultedWithIntrusiveHolders(Registries.BLOCK, "air", registry -> Blocks.AIR);
+    public static final spout.server.paper.impl.moredatadriven.minecraft.BlockRegistry BLOCK = internalRegister(Registries.BLOCK, spout.server.paper.impl.moredatadriven.minecraft.BlockRegistry.get(), registry -> Blocks.AIR); // Spout - More data-driven - Minecraft registries - Use custom class
     public static final Registry<DebugSubscription<?>> DEBUG_SUBSCRIPTION = registerSimple(Registries.DEBUG_SUBSCRIPTION, DebugSubscriptions::bootstrap);
     public static final DefaultedRegistry<EntityType<?>> ENTITY_TYPE = registerDefaultedWithIntrusiveHolders(
         Registries.ENTITY_TYPE, "pig", registry -> EntityType.PIG
     );
-    public static final DefaultedRegistry<Item> ITEM = registerDefaultedWithIntrusiveHolders(Registries.ITEM, "air", registry -> Items.AIR);
+    public static final spout.server.paper.impl.moredatadriven.minecraft.ItemRegistry ITEM = internalRegister(Registries.ITEM, spout.server.paper.impl.moredatadriven.minecraft.ItemRegistry.get(), registry -> Items.AIR); // Spout - More data-driven - Minecraft registries - Use custom class
     public static final Registry<Potion> POTION = registerSimple(Registries.POTION, Potions::bootstrap);
     public static final Registry<ParticleType<?>> PARTICLE_TYPE = registerSimple(Registries.PARTICLE_TYPE, registry -> ParticleTypes.BLOCK);
     public static final Registry<BlockEntityType<?>> BLOCK_ENTITY_TYPE = registerSimpleWithIntrusiveHolders(
@@ -367,7 +_,7 @@
     );
     // Paper end - add built-in registry conversions
 
-    private static <T> Registry<T> registerSimple(final ResourceKey<? extends Registry<T>> name, final BuiltInRegistries.RegistryBootstrap<T> loader) {
+    public static <T> Registry<T> registerSimple(final ResourceKey<? extends Registry<T>> name, final BuiltInRegistries.RegistryBootstrap<T> loader) { // Spout - Common class tweakers - Make accessible
         return internalRegister(name, new MappedRegistry<>(name, Lifecycle.stable(), false), loader);
     }
 
@@ -389,7 +_,7 @@
         return internalRegister(name, new DefaultedMappedRegistry<>(defaultKey, name, Lifecycle.stable(), true), loader);
     }
 
-    private static <T, R extends WritableRegistry<T>> R internalRegister(
+    public static <T, R extends WritableRegistry<T>> R internalRegister( // Spout - Common class tweakers - Make accessible
         final ResourceKey<? extends Registry<T>> name, final R registry, final BuiltInRegistries.RegistryBootstrap<T> loader
     ) {
         Bootstrap.checkBootstrapCalled(() -> "registry " + name.identifier());
@@ -405,6 +_,7 @@
         bootStrap(() -> {});
     }
     public static void bootStrap(Runnable runnable) {
+        spout.common.moredatadriven.minecraft.BuiltInSpoutMoreDataDrivenRegistries.bootstrap(); // Spout - More data-driven - Minecraft registries - Include Spout registries
         // Paper end
         REGISTRY.freeze(); // Paper - freeze main registry early
         createContents();
@@ -464,7 +_,7 @@
     }
 
     @FunctionalInterface
-    private interface RegistryBootstrap<T> {
+    public interface RegistryBootstrap<T> { // Spout - Common class tweakers - Make accessible
         Object run(Registry<T> registry);
     }
 }
