--- a/src/main/java/io/papermc/paper/registry/RegistryKey.java
+++ b/src/main/java/io/papermc/paper/registry/RegistryKey.java
@@ -41,6 +_,8 @@
 import org.bukkit.potion.PotionEffectType;
 import org.bukkit.potion.PotionType;
 import org.jspecify.annotations.NullMarked;
+import spout.api.gamecontent.datadriven.blocktype.BlockTypeType;
+import spout.api.gamecontent.datadriven.itemtype.ItemTypeType;
 
 import static io.papermc.paper.registry.RegistryKeyImpl.create;
 
@@ -78,6 +_,16 @@
      * @see io.papermc.paper.registry.keys.MobEffectKeys
      */
     RegistryKey<PotionEffectType> MOB_EFFECT = create("mob_effect");
+    // Spout start - More data-driven - Paper registry API - Types - API part - Define registry keys
+    /**
+     * Built-in registry for Spout block type types.
+     */
+    RegistryKey<BlockTypeType> BLOCK_TYPE_TYPE = create(spout.branding.SpoutNamespace.SPOUT + ":block_type"); // Has to match the corresponding entry in SpoutMoreDataDrivenRegistries
+    /**
+     * Built-in registry for Spout item type types.
+     */
+    RegistryKey<ItemTypeType> ITEM_TYPE_TYPE = create(spout.branding.SpoutNamespace.SPOUT + ":item_type"); // Has to match the corresponding entry in SpoutMoreDataDrivenRegistries
+    // Spout end - More data-driven - Paper registry API - Types - API part - Define registry keys
     /**
      * Built-in registry for block types.
      * @see io.papermc.paper.registry.keys.BlockTypeKeys
