--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -12,7 +_,6 @@
 import java.util.function.BiConsumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
-import java.util.function.ToIntFunction;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
@@ -83,7 +_,15 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
-public abstract class BlockBehaviour implements FeatureElement {
+// Spout start - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key
+public abstract class BlockBehaviour implements FeatureElement, spout.server.paper.impl.moredatadriven.datapack.CopyResourcesFromDataPackRegistryToInternalRegistry.DelayedResourceKeyAttachable<Block> {
+    @Override
+    public void attachResourceKey(ResourceKey<Block> resourceKey) {
+        this.properties.setId(resourceKey);
+        this.drops = properties.effectiveDrops();
+        this.descriptionId = properties.effectiveDescriptionId();
+    }
+    // Spout end - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key
     protected static final Direction[] UPDATE_SHAPE_ORDER = new Direction[]{
         Direction.WEST, Direction.EAST, Direction.NORTH, Direction.SOUTH, Direction.DOWN, Direction.UP
     };
@@ -97,13 +_,19 @@
     protected final boolean dynamicShape;
     protected final FeatureFlagSet requiredFeatures;
     protected final BlockBehaviour.Properties properties;
-    protected final Optional<ResourceKey<LootTable>> drops;
-    protected final String descriptionId;
+    // Spout start - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Make dependent fields non-final
+    protected Optional<ResourceKey<LootTable>> drops;
+    protected String descriptionId;
+    // Spout end - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Make dependent fields non-final
 
     public BlockBehaviour(final BlockBehaviour.Properties properties) {
         this.hasCollision = properties.hasCollision;
+        // Spout start - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Don't set dependent fields without a key
+        if (properties.id != null) {
         this.drops = properties.effectiveDrops();
         this.descriptionId = properties.effectiveDescriptionId();
+        }
+        // Spout end - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Don't set dependent fields without a key
         this.explosionResistance = properties.explosionResistance;
         this.isRandomlyTicking = properties.isRandomlyTicking;
         this.soundType = properties.soundType;
@@ -121,8 +_,8 @@
 
     protected abstract MapCodec<? extends Block> codec();
 
-    protected static <B extends Block> RecordCodecBuilder<B, BlockBehaviour.Properties> propertiesCodec() {
-        return BlockBehaviour.Properties.CODEC.fieldOf("properties").forGetter(BlockBehaviour::properties);
+    public static <B extends Block> RecordCodecBuilder<B, BlockBehaviour.Properties> propertiesCodec() { // Spout - More data-driven - Minecraft registries - Types - Make properties codec public
+        return spout.common.moredatadriven.minecraft.block.BlockPropertiesCodec.getBuilder(); // Spout - More data-driven - Minecraft registries - Types - Adapt existing block type registry - Make properties optional in block codec
     }
 
     public static <B extends Block> MapCodec<B> simpleCodec(final Function<BlockBehaviour.Properties, B> constructor) {
@@ -363,12 +_,20 @@
     }
 
     protected float getDestroyProgress(final BlockState state, final Player player, final BlockGetter level, final BlockPos pos) {
+        // Spout start - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
+        return this.getDestroyProgress(state, player, level, pos, null);
+    }
+
+    protected float getDestroyProgress(BlockState state, Player player, BlockGetter level, BlockPos pos, @Nullable ItemStack heldItemStack) {
+        // Spout end - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
         float destroySpeed = state.getDestroySpeed(level, pos);
         if (destroySpeed == -1.0F) {
             return 0.0F;
         } else {
-            int modifier = player.hasCorrectToolForDrops(state) ? 30 : 100;
-            return player.getDestroySpeed(state) / destroySpeed / modifier;
+            // Spout start - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
+            int i = player.hasCorrectToolForDrops(state, heldItemStack) ? 30 : 100;
+            return player.getDestroySpeed(state, heldItemStack) / destroySpeed / i;
+            // Spout end - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
         }
     }
 
@@ -452,12 +_,14 @@
         public final float destroySpeed;
         private final boolean requiresCorrectToolForDrops;
         private final boolean canOcclude;
-        private final BlockBehaviour.StatePredicate isRedstoneConductor;
-        private final BlockBehaviour.StatePredicate isSuffocating;
-        private final BlockBehaviour.StatePredicate isViewBlocking;
+        // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
+        private final spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isRedstoneConductor;
+        private final spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isSuffocating;
+        private final spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isViewBlocking;
+        // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
         private final BlockBehaviour.PostProcess postProcess;
-        private final BlockBehaviour.StatePredicate emissiveRendering;
-        private final BlockBehaviour.@Nullable OffsetFunction offsetFunction;
+        private final spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate emissiveRendering; // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
+        private final BlockBehaviour.@Nullable OffsetType offsetFunction; // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
         private final boolean spawnTerrainParticles;
         private final NoteBlockInstrument instrument;
         private final boolean replaceable;
@@ -543,7 +_,7 @@
         protected BlockStateBase(final Block owner, final Property<?>[] propertyKeys, final Comparable<?>[] propertyValues) {
             super(owner, propertyKeys, propertyValues);
             BlockBehaviour.Properties properties = owner.properties;
-            this.lightEmission = properties.lightEmission.applyAsInt(this.asState());
+            this.lightEmission = properties.lightEmission.apply(this.asState()); // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make light emission property serializable
             this.useShapeForLightOcclusion = owner.useShapeForLightOcclusion(this.asState());
             this.isAir = properties.isAir;
             this.ignitedByLava = properties.ignitedByLava;
@@ -773,7 +_,13 @@
         }
 
         public float getDestroyProgress(final Player player, final BlockGetter level, final BlockPos pos) {
-            return this.getBlock().getDestroyProgress(this.asState(), player, level, pos);
+            // Spout start - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
+            return this.getDestroyProgress(player, level, pos, null);
+        }
+
+        public float getDestroyProgress(Player player, BlockGetter level, BlockPos pos, @Nullable ItemStack heldItemStack) {
+            return this.getBlock().getDestroyProgress(this.asState(), player, level, pos, heldItemStack);
+            // Spout end - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
         }
 
         public int getDirectSignal(final BlockGetter level, final BlockPos pos, final Direction direction) {
@@ -837,8 +_,8 @@
         }
 
         public Vec3 getOffset(final BlockPos pos) {
-            BlockBehaviour.OffsetFunction function = this.offsetFunction;
-            return function != null ? function.evaluate(this.asState(), pos) : Vec3.ZERO;
+            BlockBehaviour.OffsetType offsetFunction = this.offsetFunction; // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
+            return offsetFunction != null ? offsetFunction.evaluate(this.asState(), pos) : Vec3.ZERO;
         }
 
         public boolean hasOffsetFunction() {
@@ -1093,16 +_,30 @@
         }
     }
 
-    @FunctionalInterface
-    public interface OffsetFunction {
-        Vec3 evaluate(BlockState state, BlockPos pos);
-    }
-
     public static enum OffsetType {
         NONE,
         XZ,
         XYZ;
+
+        // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
+        private final java.util.function.BiFunction<BlockState, BlockPos, Vec3> evaluate;
+
+        OffsetType() {
+            this.evaluate = BlockBehaviour.Properties.offsetFunction(this);
+        }
+
+        public Vec3 evaluate(BlockState state, BlockPos pos) {
+            return this.evaluate.apply(state, pos);
+        }
+        // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
     }
+
+    // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
+    // @FunctionalInterface
+    // public interface OffsetFunction {
+    //     Vec3 evaluate(BlockState state, BlockPos pos);
+    // }
+    // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
 
     @FunctionalInterface
     public interface PostProcess {
@@ -1110,47 +_,50 @@
     }
 
     public static class Properties {
-        public static final Codec<BlockBehaviour.Properties> CODEC = MapCodec.unitCodec(() -> of());
-        private Function<BlockState, MapColor> mapColor = state -> MapColor.NONE;
-        private boolean hasCollision = true;
-        private SoundType soundType = SoundType.STONE;
-        private ToIntFunction<BlockState> lightEmission = state -> 0;
-        private float explosionResistance;
-        private float destroyTime;
-        private boolean requiresCorrectToolForDrops;
-        private boolean isRandomlyTicking;
-        private float friction = 0.6F;
-        private float speedFactor = 1.0F;
-        private float jumpFactor = 1.0F;
-        private @Nullable ResourceKey<Block> id;
+        public static final Codec<BlockBehaviour.Properties> CODEC = spout.common.moredatadriven.minecraft.block.BlockPropertiesCodec.CODEC; // Spout - More data-driven - Minecraft registries - Block - Add missing properties codec implementation
+        // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make properties fields public
+        public spout.common.moredatadriven.minecraft.type.BlockStateFunction<MapColor> mapColor = new spout.common.moredatadriven.minecraft.type.BlockStateFunction.Single<>(MapColor.NONE);
+        public boolean hasCollision = true;
+        public SoundType soundType = SoundType.STONE;
+        public spout.common.moredatadriven.minecraft.type.BlockStateFunction<Integer> lightEmission = new spout.common.moredatadriven.minecraft.type.BlockStateFunction.Single<>(0);
+        public boolean wasExplosionResistanceSet = false; // Spout - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Replace block arguments by references - Replace stairs explosion resistance implementation
+        public float explosionResistance;
+        public float destroyTime;
+        public boolean requiresCorrectToolForDrops;
+        public boolean isRandomlyTicking;
+        public float friction = 0.6F;
+        public float speedFactor = 1.0F;
+        public float jumpFactor = 1.0F;
+        public @Nullable ResourceKey<Block> id;
         private DependantName<Block, Optional<ResourceKey<LootTable>>> drops = id -> Optional.of(
             ResourceKey.create(Registries.LOOT_TABLE, id.identifier().withPrefix("blocks/"))
         );
         private DependantName<Block, String> descriptionId = id -> Util.makeDescriptionId("block", id.identifier());
-        private boolean canOcclude = true;
-        private boolean isAir;
-        private boolean ignitedByLava;
-        @Deprecated
-        private boolean liquid;
-        @Deprecated
-        private boolean forceSolidOff;
-        private boolean forceSolidOn;
-        private PushReaction pushReaction = PushReaction.NORMAL;
-        private boolean spawnTerrainParticles = true;
-        private NoteBlockInstrument instrument = NoteBlockInstrument.HARP;
-        private boolean replaceable;
+        public boolean canOcclude = true;
+        public boolean isAir;
+        public boolean ignitedByLava;
+        @Deprecated
+        public boolean liquid;
+        @Deprecated
+        public boolean forceSolidOff;
+        public boolean forceSolidOn;
+        public PushReaction pushReaction = PushReaction.NORMAL;
+        public boolean spawnTerrainParticles = true;
+        public NoteBlockInstrument instrument = NoteBlockInstrument.HARP;
+        public boolean replaceable;
         private BlockBehaviour.StateArgumentPredicate<EntityType<?>> isValidSpawn = (state, level, pos, entityType) -> state.isFaceSturdy(
                 level, pos, Direction.UP
             )
             && state.getLightEmission() < 14;
-        private BlockBehaviour.StatePredicate isRedstoneConductor = BlockBehaviour.BlockStateBase::isCollisionShapeFullBlock;
-        private BlockBehaviour.StatePredicate isSuffocating = (state, level, pos) -> state.blocksMotion() && state.isCollisionShapeFullBlock(level, pos);
-        private BlockBehaviour.StatePredicate isViewBlocking = this.isSuffocating;
-        private BlockBehaviour.PostProcess postProcess = (state, level, pos) -> null;
-        private BlockBehaviour.StatePredicate emissiveRendering = (state, level, pos) -> false;
-        private boolean dynamicShape;
-        private FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
-        private BlockBehaviour.@Nullable OffsetFunction offsetFunction;
+        public spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isRedstoneConductor = spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.IS_COLLISION_SHAPE_FULL_BLOCK;
+        public spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isSuffocating = spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.BLOCKS_MOTION_AND_IS_COLLISION_SHAPE_FULL_BLOCK;
+        public spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isViewBlocking = this.isSuffocating;
+        public BlockBehaviour.PostProcess postProcess = (state, level, pos) -> null;
+        public spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate emissiveRendering = spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.NEVER;
+        public boolean dynamicShape;
+        public FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
+        public BlockBehaviour.@Nullable OffsetType offsetFunction;
+        // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make properties fields public
 
         private Properties() {
         }
@@ -1178,7 +_,12 @@
             BlockBehaviour.Properties copyTo = new BlockBehaviour.Properties();
             BlockBehaviour.Properties copyFrom = block.properties;
             copyTo.destroyTime = copyFrom.destroyTime;
+            // Spout start - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Replace block arguments by references - Replace stairs explosion resistance implementation
+            if (copyFrom.wasExplosionResistanceSet) {
             copyTo.explosionResistance = copyFrom.explosionResistance;
+            copyFrom.wasExplosionResistanceSet = true;
+            }
+            // Spout end - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Replace block arguments by references - Replace stairs explosion resistance implementation
             copyTo.hasCollision = copyFrom.hasCollision;
             copyTo.isRandomlyTicking = copyFrom.isRandomlyTicking;
             copyTo.lightEmission = copyFrom.lightEmission;
@@ -1204,19 +_,25 @@
             return copyTo;
         }
 
+        // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
+        public BlockBehaviour.Properties mapColor(spout.common.moredatadriven.minecraft.type.BlockStateFunction<MapColor> mapColor) {
+            this.mapColor = mapColor;
+            return this;
+        }
+        // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
+
         public BlockBehaviour.Properties mapColor(final DyeColor dyeColor) {
-            this.mapColor = state -> dyeColor.getMapColor();
-            return this;
+            return this.mapColor(dyeColor.getMapColor()); // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
         }
 
         public BlockBehaviour.Properties mapColor(final MapColor mapColor) {
-            this.mapColor = state -> mapColor;
-            return this;
+            return this.mapColor(new spout.common.moredatadriven.minecraft.type.BlockStateFunction.Single<>(mapColor)); // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
         }
 
-        public BlockBehaviour.Properties mapColor(final Function<BlockState, MapColor> mapColor) {
-            this.mapColor = mapColor;
-            return this;
+        // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
+        public BlockBehaviour.Properties mapColor(List<Property<?>> properties, Function<spout.common.moredatadriven.minecraft.type.BlockStateFunction.ByProperties.PropertyValueMap, MapColor> byPropertyValues) {
+            return this.mapColor(new spout.common.moredatadriven.minecraft.type.BlockStateFunction.ByProperties<>(properties, byPropertyValues));
+            // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make map color property serializable
         }
 
         public BlockBehaviour.Properties noCollision() {
@@ -1250,10 +_,20 @@
             return this;
         }
 
-        public BlockBehaviour.Properties lightLevel(final ToIntFunction<BlockState> lightEmission) {
-            this.lightEmission = lightEmission;
-            return this;
-        }
+        // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make light emission property serializable
+        public BlockBehaviour.Properties lightLevel(int lightEmission) {
+            return this.lightLevel(new spout.common.moredatadriven.minecraft.type.BlockStateFunction.Single<>(lightEmission));
+        }
+
+        public BlockBehaviour.Properties lightLevel(Function<BlockState, Integer> lightEmission) {
+            if (lightEmission instanceof spout.common.moredatadriven.minecraft.type.BlockStateFunction<Integer> blockStateFunction) {
+                this.lightEmission = blockStateFunction;
+                return this;
+            }
+            // Assume that every piece of code that calls this method with an argument that is not a BlockStateFunction is calling it with a single value; for every other place we call the method with a BlockStateFunction instead
+            return this.lightLevel(lightEmission.apply(null));
+        }
+        // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make light emission property serializable
 
         public BlockBehaviour.Properties strength(final float destroyTime, final float explosionResistance) {
             return this.destroyTime(destroyTime).explosionResistance(explosionResistance);
@@ -1329,16 +_,34 @@
         }
 
         public BlockBehaviour.Properties isRedstoneConductor(final BlockBehaviour.StatePredicate isRedstoneConductor) {
+            // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
+            return this.isRedstoneConductor(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.wrap(isRedstoneConductor));
+        }
+
+        public BlockBehaviour.Properties isRedstoneConductor(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isRedstoneConductor) {
+            // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
             this.isRedstoneConductor = isRedstoneConductor;
             return this;
         }
 
         public BlockBehaviour.Properties isSuffocating(final BlockBehaviour.StatePredicate isSuffocating) {
+            // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
+            return this.isSuffocating(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.wrap(isSuffocating));
+        }
+
+        public BlockBehaviour.Properties isSuffocating(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isSuffocating) {
+            // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
             this.isSuffocating = isSuffocating;
             return this;
         }
 
         public BlockBehaviour.Properties isViewBlocking(final BlockBehaviour.StatePredicate isViewBlocking) {
+            // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
+            return this.isViewBlocking(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.wrap(isViewBlocking));
+        }
+
+        public BlockBehaviour.Properties isViewBlocking(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate isViewBlocking) {
+            // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
             this.isViewBlocking = isViewBlocking;
             return this;
         }
@@ -1349,6 +_,12 @@
         }
 
         public BlockBehaviour.Properties emissiveRendering(final BlockBehaviour.StatePredicate emissiveRendering) {
+            // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
+            return this.emissiveRendering(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate.wrap(emissiveRendering));
+        }
+
+        public BlockBehaviour.Properties emissiveRendering(spout.common.moredatadriven.minecraft.common.subtypes.KnownStatePredicate emissiveRendering) {
+            // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable
             this.emissiveRendering = emissiveRendering;
             return this;
         }
@@ -1365,11 +_,19 @@
 
         public BlockBehaviour.Properties explosionResistance(final float explosionResistance) {
             this.explosionResistance = Math.max(0.0F, explosionResistance);
+            this.wasExplosionResistanceSet = true; // Spout - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Replace block arguments by references - Replace stairs explosion resistance implementation
             return this;
         }
 
         public BlockBehaviour.Properties offsetType(final BlockBehaviour.OffsetType offsetType) {
-            this.offsetFunction = switch (offsetType) {
+            // Spout start - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
+            this.offsetFunction = offsetType;
+            return this;
+        }
+
+        public static java.util.function.BiFunction<BlockState, BlockPos, Vec3> offsetFunction(BlockBehaviour.OffsetType offsetType) {
+            return switch (offsetType) {
+                // Spout end - More data-driven - Minecraft registries - Types - Properties serialization - Make offset function property serializable
                 case NONE -> null;
                 case XZ -> (state, pos) -> {
                     Block block = state.getBlock();
@@ -1389,7 +_,6 @@
                     return new Vec3(x, y, z);
                 };
             };
-            return this;
         }
 
         public BlockBehaviour.Properties noTerrainParticles() {
@@ -1433,7 +_,7 @@
     }
 
     @FunctionalInterface
-    public interface StatePredicate {
+    public interface StatePredicate extends java.io.Serializable { // Spout - More data-driven - Minecraft registries - Types - Properties serialization - Make block state predicate properties serializable - Make predicate extend serializable so that we can inspect it in KnownStatePrediate#wrap
         boolean test(BlockState state, BlockGetter level, BlockPos pos);
     }
 }
