--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -235,6 +_,7 @@
 
     private void finishLoginAndWaitForClient(final GameProfile gameProfile) {
         this.state = ServerLoginPacketListenerImpl.State.PROTOCOL_SWITCHING;
+        this.connection.clientViewSetter.sendClientModDetectionPacket(); // Spout - Client view - Set - Detect client mod
         this.connection.send(new ClientboundLoginFinishedPacket(gameProfile));
         this.server.services().paper().filledProfileCache().add(gameProfile); // Paper - update profile cache
     }
@@ -366,6 +_,7 @@
 
     @Override
     public void handleCustomQueryPacket(final ServerboundCustomQueryAnswerPacket packet) {
+        if (this.connection.clientViewSetter.handleClientModDetectionPacket(packet)) return; // Spout - Client view - Set - Handle potential client mod response
         // Paper start - Add Velocity IP Forwarding Support
         try {
             this.handleCustomQueryPacket0(packet);
