--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -102,6 +_,23 @@
     public void startConfiguration() {
         new io.papermc.paper.event.connection.configuration.PlayerConnectionInitialConfigureEvent(this.paperConnection).callEvent(); // Paper
         this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
+        // Spout start - Client view - Set - Send custom content - Send packet
+        if (this.connection.clientViewSetter.hasClientMod()) {
+            spout.clientview.clientmod.protocol.ClientModCustomContentSending.sendAllPackets(this.connection);
+        }
+        // Spout end - Client view - Set - Send custom content - Send packet
+        // Spout start - Client view - Set - Offer resource pack
+        net.minecraft.network.protocol.common.@Nullable ClientboundResourcePackPushPacket spoutResourcePackPacket = this.connection.clientViewSetter.getResourcePackPacket();
+        if (spoutResourcePackPacket != null) {
+            this.connection.clientViewSetter.isWaitingForResourcePackPacketResponse = true;
+            this.send(spoutResourcePackPacket);
+        } else {
+            this.continueConfigurationAfterOfferingSpoutResourcePack();
+        }
+    }
+
+    public void continueConfigurationAfterOfferingSpoutResourcePack() {
+        // Spout end - Client view - Set - Offer resource pack
         ServerLinks serverLinks = this.server.serverLinks();
         if (!serverLinks.isEmpty()) {
             // Paper start
@@ -165,6 +_,19 @@
         if (packet.action().isTerminal() && packet.id().equals(this.server.getServerResourcePack().map(MinecraftServer.ServerResourcePackInfo::id).orElse(null))) { // Paper - Ignore resource pack requests that are not vanilla
             this.finishCurrentTask(ServerResourcePackConfigurationTask.TYPE);
         }
+        // Spout start - Client view - Set - Offer resource pack
+        if (this.connection.clientViewSetter.isWaitingForResourcePackPacketResponse) {
+            if (packet.action() != net.minecraft.network.protocol.common.ServerboundResourcePackPacket.Action.ACCEPTED && packet.action() != net.minecraft.network.protocol.common.ServerboundResourcePackPacket.Action.DOWNLOADED) {
+                this.connection.clientViewSetter.isWaitingForResourcePackPacketResponse = false;
+                if (packet.action() == net.minecraft.network.protocol.common.ServerboundResourcePackPacket.Action.SUCCESSFULLY_LOADED) {
+                    this.connection.clientViewSetter.markAcceptedResourcePack();
+                } else {
+                    this.connection.clientViewSetter.markDeniedResourcePack();
+                }
+                this.continueConfigurationAfterOfferingSpoutResourcePack();
+            }
+        }
+        // Spout end - Client view - Set - Offer resource pack
     }
 
     @Override
