--- a/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
+++ b/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
@@ -788,6 +_,7 @@
 
         @Override
         public ServerChunkTasks queueBlockChange(final BlockPos pos) {
+
             final ServerChunkTasks ret = this.chunkTasks.compute(CoordinateUtils.getChunkKey(pos), (final long keyInMap, ServerChunkTasks valueInMap) -> {
                 if (valueInMap == null) {
                     valueInMap = new ServerChunkTasks(
@@ -798,6 +_,7 @@
                 return valueInMap;
             });
 
+            ret.requiresExtraLightSending = spout.server.paper.impl.packetmapping.block.light.CurrentUpdateRequiresExtraLightSending.get(); // Spout - Block mapping - Update block light for custom blocks
             ret.schedule();
 
             return ret;
@@ -817,6 +_,7 @@
                 return valueInMap;
             });
 
+            ret.requiresExtraLightSending = spout.server.paper.impl.packetmapping.block.light.CurrentUpdateRequiresExtraLightSending.get(); // Spout - Block mapping - Update block light for custom blocks
             ret.schedule();
 
             return ret;
@@ -882,6 +_,7 @@
 
             private final AtomicBoolean ticketAdded = new AtomicBoolean();
             private final PrioritisedExecutor.PrioritisedTask task;
+            public boolean requiresExtraLightSending; // Spout - Block mapping - Update block light for custom blocks
 
             public ServerChunkTasks(final long chunkCoordinate, final StarLightInterface lightEngine,
                                     final ServerLightQueue queue) {
@@ -929,7 +_,9 @@
             public void run() {
                 ((ServerLightQueue)this.queue).chunkTasks.remove(this.chunkCoordinate, this);
 
+                if (this.requiresExtraLightSending) spout.server.paper.impl.packetmapping.block.light.CurrentTaskRequiresExtraLightSending.set(); // Spout - Block mapping - Update block light for custom blocks
                 this.runTasks();
+                if (this.requiresExtraLightSending) spout.server.paper.impl.packetmapping.block.light.CurrentTaskRequiresExtraLightSending.clear(); // Spout - Block mapping - Update block light for custom blocks
             }
         }
     }
