--- a/paper-api/build.gradle.kts
+++ b/paper-api/build.gradle.kts
@@ -91,7 +_,8 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher:6.0.3")
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path =
+    rootProject.layout.projectDirectory.dir("paper-api/src/generated/java").asFile.toPath()
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -101,6 +_,26 @@
     main {
         java {
             srcDir(generatedDir)
+            // Spout start - Project setup - Set project directory references
+            srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../../common-src/src/main/java"))
+            srcDir(file("src/not-api/java"))
+            // Spout end - Project setup - Set project directory references
+        }
+        resources {
+            // Spout start - Project setup - Set project directory references
+            srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../../common-src/src/main/resources"))
+            srcDir(file("src/not-api/resources"))
+            // Spout end - Project setup - Set project directory references
+        }
+    }
+    test {
+        java {
+            srcDir(file("../paper-api/src/test/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/test/resources"))
         }
     }
 }
@@ -116,10 +_,27 @@
             }
             outgoing {
                 capability(mainCapability)
+                // Spout start - Project setup - Extend module capabilities
+                // Conflict with jitpack.io artifact requirements
+                fun getGitHashProvider(length: Int?) = providers.exec {
+                    workingDir = rootProject.projectDir
+                    commandLine(*(listOfNotNull("git", "rev-parse", length?.let { "--short=$it" }, "HEAD").toTypedArray()))
+                }.standardOutput.asText.map { it.trim() }
+                for (coordinates in arrayOf(
+                    "com.github.modernspout:spout-api",
+                    "com.github.ModernSpout:Spout"
+                )) {
+                    for (version in arrayOf(project.version, getGitHashProvider(null).get(), getGitHashProvider(7).get(), getGitHashProvider(8).get(), getGitHashProvider(10).get())) {
+                        val alternativeCapability = "$coordinates:$version"
+                        if (alternativeCapability != mainCapability) capability(alternativeCapability)
+                    }
+                }
+                // Spout end - Project setup - Extend module capabilities
                 // Paper-MojangAPI has been merged into Paper-API
                 capability("io.papermc.paper:paper-mojangapi:${project.version}")
                 capability("com.destroystokyo.paper:paper-mojangapi:${project.version}")
                 // Conflict with old coordinates
+                capability("io.papermc.paper:paper-api:${project.version}") // Spout - Project setup - Extend module capabilities
                 capability("com.destroystokyo.paper:paper-api:${project.version}")
                 capability("org.spigotmc:spigot-api:${project.version}")
                 capability("org.bukkit:bukkit:${project.version}")
@@ -179,11 +_,18 @@
     @get:Inject
     abstract val fileSystemOperations: FileSystemOperations
 }
+
 val services = objects.newInstance<Services>()
 
 tasks.withType<Javadoc>().configureEach {
+    // Spout start - Project setup - Hide Javadoc lint warnings
+    (options as StandardJavadocDocletOptions).addStringOption(
+        "Xdoclint:none",
+        "-quiet"
+    )
+    // Spout end - Project setup - Hide Javadoc lint warnings
     val options = options as StandardJavadocDocletOptions
-    options.overview = "src/main/javadoc/overview.html"
+    options.overview = "../paper-api/src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
@@ -212,16 +_,18 @@
     doFirst {
         options.addStringOption(
             "sourcepath",
-            apiAndDocsElements.get().map { it.asFile }.joinToString(separator = File.pathSeparator, transform = File::getPath)
+            apiAndDocsElements.get().map { it.asFile }
+                .joinToString(separator = File.pathSeparator, transform = File::getPath)
         )
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
-    inputs.dir("src/main/javadoc").withPropertyName("javadoc-sourceset")
+    inputs.dir("../spout-api/src/main/javadoc")
+        .withPropertyName("javadoc-sourceset") // Spout start - Project setup - Update fix to fork directory
     val fsOps = services.fileSystemOperations
     doLast {
         fsOps.copy {
-            from("src/main/javadoc") {
+            from("../spout-api/src/main/javadoc") { // Spout start - Project setup - Update fix to fork directory
                 include("**/doc-files/**")
             }
             into("build/docs/javadoc")
@@ -251,3 +_,16 @@
 tasks.check {
     dependsOn(scanJarForBadCalls)
 }
+
+// Spout start - Project setup - Hide annoying compilation warnings
+tasks.withType<JavaCompile> {
+    options.compilerArgs.addAll(
+        listOf(
+            "-Xlint:-dep-ann",
+            "-Xlint:-deprecation",
+            "-Xlint:-module",
+            "-Xlint:-removal",
+        )
+    )
+}
+// Spout end - Project setup - Hide annoying compilation warnings
