--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -361,8 +_,10 @@
     }
 
     @Override
-    public Packet<ClientGamePacketListener> getAddEntityPacket(final ServerEntity serverEntity) {
-        return new ClientboundAddEntityPacket(this, serverEntity, Block.getId(this.getBlockState()));
+    // Spout start - Block mapping - Implementation part - Apply to entity block states
+    public Packet<ClientGamePacketListener> getAddEntityPacket(final ServerEntity serverEntity, net.minecraft.server.level.@Nullable ServerPlayer player) {
+        return new ClientboundAddEntityPacket(this, serverEntity, spout.server.paper.impl.packetmapping.block.MappedBlockStateRegistry.get().getIdNullable(this.getBlockState(), player));
+        // Spout end - Block mapping - Implementation part - Apply to entity block states
     }
 
     @Override
