--- a/net/minecraft/network/protocol/common/ClientboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/common/ClientboundCustomPayloadPacket.java
@@ -20,7 +_,7 @@
         )
         .map(ClientboundCustomPayloadPacket::new, ClientboundCustomPayloadPacket::payload);
     public static final StreamCodec<FriendlyByteBuf, ClientboundCustomPayloadPacket> CONFIG_STREAM_CODEC = CustomPacketPayload.<FriendlyByteBuf>codec(
-            id -> DiscardedPayload.codec(id, 1048576), List.of(new CustomPacketPayload.TypeAndCodec<>(BrandPayload.TYPE, BrandPayload.STREAM_CODEC))
+            id -> DiscardedPayload.codec(id, 1048576), List.of(new CustomPacketPayload.TypeAndCodec<>(BrandPayload.TYPE, BrandPayload.STREAM_CODEC), spout.clientview.clientmod.protocol.ClientModCustomContentPacketPayload.TYPE_AND_CODEC) // Spout - Client view - Set - Send custom content - Send packet - Add codec for packet
         )
         .map(ClientboundCustomPayloadPacket::new, ClientboundCustomPayloadPacket::payload);
 
