--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -233,10 +_,15 @@
         }
     }
 
-    public record DataValue<T>(int id, EntityDataSerializer<T> serializer, T value) {
+    // Spout start - Item mapping - Detect item frame - Store accessor in DataValue
+    public record DataValue<T>(@Nullable EntityDataAccessor<T> dataAccessor, int id, EntityDataSerializer<T> serializer, T value) {
+        DataValue(int id, EntityDataSerializer<T> serializer, T value) {
+            this(null, id, serializer, value);
+        }
+        // Spout end - Item mapping - Detect item frame - Store accessor in DataValue
         public static <T> SynchedEntityData.DataValue<T> create(final EntityDataAccessor<T> accessor, final T value) {
             EntityDataSerializer<T> serializer = accessor.serializer();
-            return new SynchedEntityData.DataValue<>(accessor.id(), serializer, serializer.copy(value));
+            return new SynchedEntityData.DataValue<>(accessor, accessor.id(), serializer, serializer.copy(value)); // Spout - Item mapping - Detect item frame - Store accessor in DataValue
         }
 
         public void write(final RegistryFriendlyByteBuf output) {
@@ -247,7 +_,12 @@
 
             output.writeByte(this.id);
             output.writeVarInt(serializerId);
+            // Spout start - Item mapping - Detect item frame - Set flag if item frame
+            boolean isItemFrame = this.dataAccessor == net.minecraft.world.entity.decoration.ItemFrame.DATA_ITEM;
+            if (isItemFrame) output.isEncodingItemFrameItem = true;
+            // Spout end - Item mapping - Detect item frame - Set flag if item frame
             this.serializer.codec().encode(output, this.value);
+            if (isItemFrame) output.isEncodingItemFrameItem = false; // Spout - Item mapping - Detect item frame - Set flag if item frame
         }
 
         public static SynchedEntityData.DataValue<?> read(final RegistryFriendlyByteBuf input, final int id) {
