--- a/src/main/java/io/papermc/paper/plugin/entrypoint/LaunchEntryPointHandler.java
+++ b/src/main/java/io/papermc/paper/plugin/entrypoint/LaunchEntryPointHandler.java
@@ -27,6 +_,7 @@
     // Utility
     public static void enterBootstrappers() {
         LaunchEntryPointHandler.INSTANCE.enter(Entrypoint.BOOTSTRAPPER);
+        spout.server.paper.impl.resourcepack.plugin.discover.PluginResourcePackDiscoveryImpl.get().fireComposeEvent(); // Spout - Resource pack - Plugin resource pack discovery - Implementation part - Fire event
     }
 
     @Override
