--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -123,7 +_,7 @@
 import net.minecraft.world.scores.Team;
 import org.jspecify.annotations.Nullable;
 
-public abstract class Player extends Avatar implements ContainerUser {
+public abstract class Player extends Avatar implements ContainerUser, spout.server.paper.impl.clientview.lookup.ClientViewLookup { // Spout - Client view - Lookup - Provide from Player
     public static final int MAX_HEALTH = 20;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
@@ -622,7 +_,13 @@
     }
 
     public float getDestroySpeed(final BlockState state) {
-        float speed = this.inventory.getSelectedItem().getDestroySpeed(state);
+        // Spout start - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
+        return this.getDestroySpeed(state, null);
+    }
+
+    public float getDestroySpeed(BlockState state, @Nullable ItemStack heldItemStack) {
+        float speed = (heldItemStack != null ? heldItemStack : this.inventory.getSelectedItem()).getDestroySpeed(state);
+        // Spout end - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
         if (speed > 1.0F) {
             speed += (float)this.getAttributeValue(Attributes.MINING_EFFICIENCY);
         }
@@ -654,7 +_,13 @@
     }
 
     public boolean hasCorrectToolForDrops(final BlockState state) {
-        return !state.requiresCorrectToolForDrops() || this.inventory.getSelectedItem().isCorrectToolForDrops(state);
+        // Spout start - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
+        return this.hasCorrectToolForDrops(state, null);
+    }
+
+    public boolean hasCorrectToolForDrops(BlockState state, @Nullable ItemStack heldItemStack) {
+        return !state.requiresCorrectToolForDrops() || (heldItemStack != null ? heldItemStack : this.inventory.getSelectedItem()).isCorrectToolForDrops(state);
+        // Spout end - Block mapping - Breaking speed - Implement calculating destroy progress with given held item stack
     }
 
     @Override
