--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -330,7 +_,7 @@
                 .getOrThrow(net.minecraft.world.level.biome.Biomes.PLAINS);
             player.connection.send(new net.minecraft.network.protocol.game.ClientboundLevelChunkWithLightPacket(
                 new net.minecraft.world.level.chunk.EmptyLevelChunk(level, player.chunkPosition(), plains),
-                level.getLightEngine(), (java.util.BitSet)null, (java.util.BitSet) null, true) // Paper - Anti-Xray
+                level.getLightEngine(), (java.util.BitSet)null, (java.util.BitSet) null, true, player) // Paper - Anti-Xray // Spout - Block mapping - Chunk packets - Pass the player
             );
         }
         // Paper end - Send empty chunk
@@ -742,6 +_,7 @@
             // CraftBukkit end
             this.sendAllPlayerInfoIn = 0;
         }
+        spout.server.paper.impl.packetmapping.block.breakspeed.BlockBreakSpeedFactorUpdater.updateFactor(this.players); // Spout - Block mapping - Breaking speed - Update factor
     }
 
     // CraftBukkit start - add a world/entity limited version
@@ -1018,7 +_,7 @@
         // needs to be done because the ServerPlayer instance is being reused on respawn instead of getting replaced like on vanilla
         java.util.Collection<net.minecraft.world.entity.ai.attributes.AttributeInstance> syncableAttributes = player.getAttributes().getSyncableAttributes();
         player.getBukkitEntity().injectScaledMaxHealth(syncableAttributes, true);
-        player.connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket(player.getId(), syncableAttributes));
+        player.connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket(player.getId(), syncableAttributes, player)); // Spout - Block mapping - Breaking speed - Pass self player to attributes packet
         // Paper end - send all attributes
         player.refreshEntityData(player); // CraftBukkit - SPIGOT-7218: sync metadata
         player.connection.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
