--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -29,6 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public class BlockItem extends Item {
+    @Override public spout.gamecontent.datadriven.itemtype.SpoutItemType spout$getItemType() { return spout.gamecontent.datadriven.itemtype.SpoutItemTypes.BLOCK; } // Spout - More data-driven - Minecraft registries - Block - Implement interfaces
     @Deprecated
     private final Block block;
 
