--- a/net/minecraft/world/item/DoubleHighBlockItem.java
+++ b/net/minecraft/world/item/DoubleHighBlockItem.java
@@ -8,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class DoubleHighBlockItem extends BlockItem {
+    @Override public spout.gamecontent.datadriven.itemtype.SpoutItemType spout$getItemType() { return spout.gamecontent.datadriven.itemtype.SpoutItemTypes.DOUBLE_HIGH_BLOCK; } // Spout - More data-driven - Minecraft registries - Block - Implement interfaces
+
     public DoubleHighBlockItem(final Block block, final Item.Properties properties) {
         super(block, properties);
     }
