--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -95,16 +_,50 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class Item implements ItemLike, FeatureElement {
+// Spout start - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key
+// Spout start - More data-driven - Minecraft registries - Block - Implement interfaces
+public class Item implements FeatureElement, ItemLike, spout.common.moredatadriven.minecraft.item.ItemTypeDecorator, spout.server.paper.impl.moredatadriven.datapack.CopyResourcesFromDataPackRegistryToInternalRegistry.DelayedResourceKeyAttachable<Item> {
+    @Override public spout.common.moredatadriven.minecraft.itemtype.SpoutItemType spout$getItemType() { return spout.common.moredatadriven.minecraft.itemtype.SpoutItemTypes.ITEM; }
+    // Spout end - More data-driven - Minecraft registries - Block - Implement interfaces
+    private Item.@Nullable Properties originalProperties;
+    @Override
+    public void attachResourceKey(ResourceKey<Item> resourceKey) {
+        Item.Properties properties = this.originalProperties;
+        this.originalProperties = null;
+        properties.setId(resourceKey);
+        this.descriptionId = properties.effectiveDescriptionId();
+        DataComponentInitializers.Initializer<Item> componentInitializer = properties.finalizeInitializer(
+            Component.translatable(this.descriptionId), properties.effectiveModel()
+        );
+        BuiltInRegistries.DATA_COMPONENT_INITIALIZERS.add(properties.itemIdOrThrow(), componentInitializer);
+    }
+    // Spout end - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key
     public static final Codec<Holder<Item>> CODEC = BuiltInRegistries.ITEM
         .holderByNameCodec()
         .validate(item -> item.is(Items.AIR.builtInRegistryHolder()) ? DataResult.error(() -> "Item must not be minecraft:air") : DataResult.success(item));
-    public static final StreamCodec<RegistryFriendlyByteBuf, Holder<Item>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.ITEM);
     public static final Codec<Holder<Item>> CODEC_WITH_BOUND_COMPONENTS = CODEC.validate(
         item -> !item.areComponentsBound()
             ? DataResult.error(() -> "Item " + item.getRegisteredName() + " does not have components yet")
             : DataResult.success(item)
     );
+    // Spout start - Item mapping - Wrap codec for items
+    public static final StreamCodec<RegistryFriendlyByteBuf, Holder<Item>> STREAM_CODEC = new net.minecraft.network.codec.StreamCodec<>() {
+
+        @Override
+        public Holder<Item> decode(final RegistryFriendlyByteBuf buffer) {
+            return UNMAPPED_STREAM_CODEC.decode(buffer);
+        }
+
+        @Override
+        public void encode(final RegistryFriendlyByteBuf buffer, final Holder<Item> value) {
+            Item item = value.value();
+            Item mappedItem = spout.server.paper.impl.packetmapping.item.ItemMappingsImpl.get().applyGenerically(item);
+            UNMAPPED_STREAM_CODEC.encode(buffer, mappedItem == item ? value : Holder.direct(mappedItem));
+        }
+
+    };
+    public static final StreamCodec<RegistryFriendlyByteBuf, Holder<Item>> UNMAPPED_STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.ITEM);
+    // Spout end - Item mapping - Wrap codec for items
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final Map<Block, Item> BY_BLOCK = Maps.newHashMap();
     public static final Identifier BASE_ATTACK_DAMAGE_ID = Identifier.withDefaultNamespace("base_attack_damage");
@@ -114,14 +_,82 @@
     public static final int MAX_BAR_WIDTH = 13;
     protected static final int APPROXIMATELY_INFINITE_USE_DURATION = 72000;
     private final Holder.Reference<Item> builtInRegistryHolder = BuiltInRegistries.ITEM.createIntrusiveHolder(this);
+    private DataComponentMap components; // Spout - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Make dependent fields non-final
     private final @Nullable ItemStackTemplate craftingRemainingItem;
-    protected final String descriptionId;
+    protected String descriptionId; // Spout - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Make dependent fields non-final
     private final FeatureFlagSet requiredFeatures;
 
     public static int getId(final Item item) {
         return item == null ? 0 : BuiltInRegistries.ITEM.getId(item);
     }
 
+    public java.util.@Nullable List<spout.server.paper.impl.packetmapping.item.datadriven.UnappliedDataDrivenItemMapping> unappliedDataPackMappings = null; // Spout - Item mapping - Data-driven - Load mappings from data pack
+
+    // Spout start - More data-driven - Minecraft registries - Cache block and item values
+    /**
+     * The {@link net.minecraft.resources.Identifier} with which this item is registered in
+     * {@link BuiltInRegistries#ITEM}.
+     *
+     * <p>
+     * This will be null until initialized in
+     * {@link spout.server.paper.impl.moredatadriven.minecraft.ItemRegistry#register}.
+     * </p>
+     */
+    public net.minecraft.resources.Identifier keyInItemRegistry = null;
+
+    /**
+     * The {@linkplain net.minecraft.core.Registry#getId index} at which this item is registered in
+     * {@link BuiltInRegistries#ITEM}.
+     *
+     * <p>
+     * This will be -1 until initialized in
+     * {@link spout.server.paper.impl.moredatadriven.minecraft.ItemRegistry#register}.
+     * </p>
+     */
+    public int indexInItemRegistry = -1;
+
+    /**
+     * The {@linkplain net.minecraft.core.Registry#getId index} at which this item is registered in
+     * {@link spout.server.paper.impl.moredatadriven.minecraft.VanillaOnlyItemRegistry}.
+     *
+     * <p>
+     * This will be -1 until initialized in
+     * {@link spout.server.paper.impl.moredatadriven.minecraft.VanillaOnlyItemRegistry#add}.
+     * </p>
+     */
+    public int indexInVanillaOnlyItemRegistry = -1;
+    // Spout end - More data-driven - Minecraft registries - Cache block and item values
+
+    // Spout start - More data-driven - Vanilla flag - Implementation part
+    private boolean isVanilla;
+
+    /**
+     * Returns whether this item is present in vanilla.
+     *
+     * <p>
+     * This does not take into account whether any details of this item, such as its max stack size,
+     * are different (were modified) from vanilla.
+     * </p>
+     *
+     * @return Whether this item is present in vanilla.
+     */
+    public boolean isVanilla() {
+        return this.isVanilla;
+    }
+
+    /**
+     * Sets {@link #isVanilla} to true.
+     *
+     * <p>
+     * This must only be called at most once,
+     * before the item is registered to {@link BuiltInRegistries#ITEM}.
+     * </p>
+     */
+    public void markAsVanilla() {
+        this.isVanilla = true;
+    }
+    // Spout end - More data-driven - Vanilla flag - Implementation part
+
     public static Item byId(final int id) {
         return BuiltInRegistries.ITEM.byId(id);
     }
@@ -132,11 +_,17 @@
     }
 
     public Item(final Item.Properties properties) {
+        // Spout start - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Don't set dependent fields without a key
+        if (properties.id != null) {
         this.descriptionId = properties.effectiveDescriptionId();
         DataComponentInitializers.Initializer<Item> componentInitializer = properties.finalizeInitializer(
             Component.translatable(this.descriptionId), properties.effectiveModel()
         );
         BuiltInRegistries.DATA_COMPONENT_INITIALIZERS.add(properties.itemIdOrThrow(), componentInitializer);
+        } else {
+            this.originalProperties = properties;
+        }
+        // Spout end - More data-driven - Data pack - Copy loaded blocks and items to their internal registries - Allow delayed setting of key - Don't set dependent fields without a key
         this.craftingRemainingItem = properties.craftingRemainingItem;
         this.requiredFeatures = properties.requiredFeatures;
         if (SharedConstants.IS_RUNNING_IN_IDE) {
@@ -373,15 +_,20 @@
         return false;
     }
 
-    public static class Properties {
+    // Spout start - More data-driven - Minecraft registries - Item - Implement interfaces
+    public static class Properties implements spout.common.moredatadriven.minecraft.item.ItemPropertiesSerializationDecorator {
+        private @Nullable DataComponentMap initializedComponents = DataComponentMap.EMPTY;
+        @Override public @Nullable DataComponentMap spout$initializedComponents() { return this.initializedComponents; }
+        @Override public void spout$initializedComponents(@Nullable DataComponentMap initializedComponents) { this.initializedComponents = initializedComponents; }
+        // Spout end - More data-driven - Minecraft registries - Item - Implement interfaces
         private static final DependantName<Item, String> BLOCK_DESCRIPTION_ID = id -> Util.makeDescriptionId("block", id.identifier());
         private static final DependantName<Item, String> ITEM_DESCRIPTION_ID = id -> Util.makeDescriptionId("item", id.identifier());
-        private DataComponentInitializers.Initializer<Item> componentInitializer = (builder, context, id) -> builder.addAll(
+        public DataComponentInitializers.Initializer<Item> componentInitializer = (builder, context, id) -> builder.addAll( // Spout - Common class tweakers - Make accessible
             DataComponents.COMMON_ITEM_COMPONENTS
         );
         private @Nullable ItemStackTemplate craftingRemainingItem;
-        private FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
-        private @Nullable ResourceKey<Item> id;
+        public FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET; // Spout - Common class tweakers - Make accessible
+        public @Nullable ResourceKey<Item> id; // Spout - Common class tweakers - Make accessible
         private DependantName<Item, String> descriptionId = ITEM_DESCRIPTION_ID;
         private final DependantName<Item, Identifier> model = ResourceKey::identifier;
 
