--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -112,8 +_,14 @@
             start(Registries.GAME_EVENT, RegistryKey.GAME_EVENT).craft(GameEvent.class, CraftGameEvent::new).writable(PaperGameEventRegistryEntry.PaperBuilder::new),
             start(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE).craft(StructureType.class, CraftStructureType::new).build(),
             start(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT).craft(PotionEffectType.class, CraftPotionEffectType::new).build(),
-            start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).build(),
-            start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).build(),
+            // Spout start - More data-driven - Paper registry API - Types - Implementation part - Create Paper registries
+            start(spout.common.moredatadriven.minecraft.SpoutMoreDataDrivenRegistries.BLOCK_TYPE, io.papermc.paper.registry.RegistryKey.BLOCK_TYPE_TYPE).craft(spout.common.moredatadriven.minecraft.blocktype.SpoutBlockType.class, spout.server.paper.impl.moredatadriven.paper.registry.type.BlockTypeTypeImpl::of).writable(spout.server.paper.impl.moredatadriven.paper.registry.type.BlockTypeTypeRegistryEntryImpl.BuilderImpl::new).delayed(),
+            start(spout.common.moredatadriven.minecraft.SpoutMoreDataDrivenRegistries.ITEM_TYPE, io.papermc.paper.registry.RegistryKey.ITEM_TYPE_TYPE).craft(spout.common.moredatadriven.minecraft.itemtype.SpoutItemType.class, spout.server.paper.impl.moredatadriven.paper.registry.type.ItemTypeTypeImpl::of).writable(spout.server.paper.impl.moredatadriven.paper.registry.type.ItemTypeTypeRegistryEntryImpl.BuilderImpl::new).delayed(),
+            // Spout end - More data-driven - Paper registry API - Types - Implementation part - Create Paper registries
+            // Spout start - More data-driven - Paper registry API - Implementation part - Make writable and delayed
+            start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).writable(spout.server.paper.impl.moredatadriven.paper.registry.BlockTypeRegistryEntryImpl.BuilderImpl::new).delayed(),
+            start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).writable(spout.server.paper.impl.moredatadriven.paper.registry.ItemTypeRegistryEntryImpl.BuilderImpl::new).delayed(),
+            // Spout end - More data-driven - Paper registry API - Implementation part - Make writable and delayed
             start(Registries.VILLAGER_PROFESSION, RegistryKey.VILLAGER_PROFESSION).craft(Villager.Profession.class, CraftVillager.CraftProfession::new).build(),
             start(Registries.VILLAGER_TYPE, RegistryKey.VILLAGER_TYPE).craft(Villager.Type.class, CraftVillager.CraftType::new).build(),
             start(Registries.MAP_DECORATION_TYPE, RegistryKey.MAP_DECORATION_TYPE).craft(MapCursor.Type.class, CraftMapCursor.CraftType::new).build(),
