--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -896,7 +_,7 @@
             Preconditions.checkArgument(effect.getData() == null || effect == Effect.ELECTRIC_SPARK, "Wrong kind of data for the %s effect", effect);
         }
 
-        int datavalue = CraftEffect.getDataValue(effect, data);
+        Object datavalue = CraftEffect.getDataValue(effect, data); // Spout - Packet mapping - Implementation part - Apply to level events
         this.playEffect(loc, effect, datavalue);
     }
 
@@ -2578,7 +_,7 @@
     public void updateScaledHealth(boolean sendHealth) {
         // SPIGOT-3813: Attributes before health
         if (this.getHandle().connection != null) {
-            this.getHandle().connection.send(new ClientboundUpdateAttributesPacket(this.getHandle().getId(), Set.of(this.getScaledMaxHealth())));
+            this.getHandle().connection.send(new ClientboundUpdateAttributesPacket(this.getHandle().getId(), Set.of(this.getScaledMaxHealth()), null)); // Spout - Block mapping - Breaking speed - Pass self player to attributes packet
             if (sendHealth) {
                 this.sendHealthUpdate();
             }
